### ISSUE TITLE
Issue building SDK for iOS (compile erros in /boost/geometry)

### ISSUE THREAD

**Issue by asdafa3** [CartoDB/mobile-sdk#421]
**Problem:**

I can't get the carto sdk to build for iOS on MacOS Big Sur, following the setup instructions in the readme almost one to one.
Running the build with 
```
python build-ios.py --profile standard
```
two types of errors are thrown (compiling libs-external/boost/geometry files):
1. error: constexpr function's return type 'void' is not a literal type
2. no template named '<template_name>_t' in namespace 'std'; did you mean simply '<template_name>_t'

Both types of errors stem from the wrong c++ compiler version being picked (I guess at least the first language feature is a feature of c++14).

**What I did before running the build script:**

Since the proposed 
```
git submodule update --init --remote --recursive
```
was not working properly with errors 
```
fatal: Submodule (SUB_PROJECT) branch configured to inherit branch from 
superproject, but the superproject is not on any branch
fatal: Needed a single revision
Unable to find current origin/ revision in submodule path 'SUB_PROJECT'
```
I used the following workaround script
```
#!/bin/bash

function prepare_submodules {
	git submodule update --init
	local SUBMODULE_LIST=`git submodule--helper status | awk '{print $2}'`
	while read -r submod; do
		local PARENT_DIR=`pwd`
		local BRANCH=`git submodule--helper remote-branch ${submod}`

		cd ${submod}
		git branch -f ${BRANCH} HEAD;
		git checkout ${BRANCH}
		if [ ! -z "`git submodule status`" ]; then
			prepare_submodules;
		fi
		cd ${PARENT_DIR}
	done <<< "$SUBMODULE_LIST"
}
```
along with
```
$ git submodule update --rebase --recursive --remote
```
as proposed [here](https://public-inbox.org/git/08844c07-5192-0645-9bfc-62e1b16e39a8@samsung.com/T/) and up to now it worked fine, no errors thrown in git submodule command.

Setting up the boost library as follows
```
cd libs-external/boost
./bootstrap.sh
./b2 headers
cd ../..
```
works well again with no errors.

Also the swig installation and setup worked just as expected.

As I thought of it being a compiler version problem, I tried various methods to call the build script with --cmake-options and tried to force the compiler version choice for cmake from here but this seemed pointless. Nonetheless, clang 12 has support for all these c++ versions and the build scripts should actually choose the right compiler version as far as I see reading through the config.hpp and CMake files, so don't really understand the root of this issue.

It would be nice if someone could give me a hint on whether this problem stems from my side (maybe I miss something fundamental) or is actually an issue in the build script.

**Additional Information:**

Branch: master, HEAD: 060d77b7
Compiler: Apple clang version 12.0.0 (clang-1200.0.32.6)
CMake: 3.19.3
XCode: 12.4 (12D4e)
Git: 2.24.3 (Apple Git-128)
Python: 3.9.1

**Comment by mtehver** [comment]
This is specific to boost 1.75+. Starting from this version boost-geometry requires C++14 compiler (or compilation mode), while the SDK uses C++11. I suggest using boost 1.74 or 1.73.

**Comment by asdafa3** [comment]
I am very grateful for this fast response, but I guess I have to ask again. I followed your suggestion and installed boost 1.74 on my system. I also verified the version and so far it seems to be correct. But unfortunately, the build errors remain unchanged. I assumed, that maybe in the build process, the boost version from the submodule may be build again and could somehow be written back, but this is not the case (and seems fairly unreasonable). Do you have any thoughts on this?

**Comment by asdafa3** [comment]
Nonetheless, I am afraid I have to ask something again. I now tried to install boost 1.74 on my system and explicitly set the compiler to c++11 when using b2, just to be sure. I totally removed the boost submodule from my local project, but this way it fails because files are not found. So I went into the build scripts and went into /build/CMakeLists.txt and tried to set the include directory for boost to my local Boost include directory (default /usr/local/include/boost). But it always complains header files not found (although they are there, so the path is not chosen). What does not make any sense for header-only files in this case, right? Further I am concerned about linking, at least I am unable to figure out where the actual linking of non header-only library files is done in this project. It thought it would have to be done in some cmake configuration file, but I can't find anything helpful. 

Another question: Is this at least the way to go in my case or is there a way to tell the submodule to use another commit or a tag possibly from version 1.74 or before?

**Comment by mtehver** [comment]
It is difficult to suggest anything without seeing actual configuration. A simple hack-fix would be to manually download 1.74 as .zip from boost homepage, unzip it under 'libs-external' and rename 'boost_1_74_0' to 'boost'. Running b2 or bjam is not required in this case. Regarding linking to boost libraries, this is not needed as SDK only uses 'header-only' parts of the boost library.

**Comment by asdafa3** [comment]
I got it resolved now, thanks.
